File: smarttrade-web/src/features/trading-journal/data/repositories/assetPreferenceRepositoryImpl.ts
import { DatabaseFailure } from '@/core/error/failure'
import { err, ok, type Result } from '@/core/result/result'
import type { AssetPreference } from '../../domain/entities/assetPreference'
import type { AssetPreferenceRepository } from '../../domain/repositories/assetPreferenceRepository'
import { assetPreferenceLocalDataSource } from '../datasources/assetPreferenceLocalDataSource'

export class AssetPreferenceRepositoryImpl implements AssetPreferenceRepository {
  async getAll(): Promise<Result<AssetPreference[]>> {
    try {
      return ok(await assetPreferenceLocalDataSource.getAll())
    } catch (error) {
      return err(new DatabaseFailure(`Falha ao carregar preferências: ${error}`))
    }
  }

  async save(preference: AssetPreference): Promise<Result<void>> {
    try {
      await assetPreferenceLocalDataSource.save(preference)
      return ok(undefined)
    } catch (error) {
      return err(new DatabaseFailure(`Falha ao salvar preferência: ${error}`))
    }
  }
}

export const assetPreferenceRepository: AssetPreferenceRepository =
  new AssetPreferenceRepositoryImpl()

File: smarttrade_app/lib/features/trading_journal/data/repositories/asset_preference_repository_impl.dart
import 'package:dartz/dartz.dart';
import 'package:smarttrade_app/core/error/failures.dart';
import 'package:smarttrade_app/features/trading_journal/data/datasources/asset_preference_local_datasource.dart';
import 'package:smarttrade_app/features/trading_journal/domain/entities/asset_preference.dart';
import 'package:smarttrade_app/features/trading_journal/domain/repositories/asset_preference_repository.dart';

class AssetPreferenceRepositoryImpl implements AssetPreferenceRepository {
  const AssetPreferenceRepositoryImpl(this._localDataSource);

  final AssetPreferenceLocalDataSource _localDataSource;

  @override
  Future<Either<Failure, List<AssetPreference>>> getAll() async {
    try {
      return Right(await _localDataSource.getAll());
    } catch (error) {
      return Left(DatabaseFailure('Falha ao carregar preferências: $error'));
    }
  }

  @override
  Future<Either<Failure, Unit>> save(AssetPreference preference) async {
    try {
      await _localDataSource.save(preference);
      return const Right(unit);
    } catch (error) {
      return Left(DatabaseFailure('Falha ao salvar preferência: $error'));
    }
  }
}
